src/orange/BoundingBoxIO.json.cc

```diff
@@ -0,0 +1,120 @@
+//! \file orange/BoundingBoxIO.json.cc
+//---------------------------------------------------------------------------//
+#include "BoundingBoxIO.json.hh"
+
+#include <cmath>
+#include <limits>
+
+#include "corecel/cont/ArrayIO.json.hh"
+#include "corecel/cont/Range.hh"
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+//! Replace "max" with "inf" since the latter can't be represented in JSON.
+template<class T>
+void max_to_inf(celeritas::Array<T, 3>* point)
+{
+    constexpr auto max_real = std::numeric_limits<T>::max();
+    constexpr auto inf = std::numeric_limits<T>::infinity();
+
+    for (auto axis : range(celeritas::Axis::size_))
+    {
+        auto ax = to_int(axis);
+        if (std::fabs((*point)[ax]) == max_real)
+        {
+            (*point)[ax] = std::copysign(inf, (*point)[ax]);
+        }
+    }
+}
+
+//---------------------------------------------------------------------------//
+//! Replace "inf" with "max" since the former can't be represented in JSON.
+template<class T>
+void inf_to_max(celeritas::Array<T, 3>* point)
+{
+    constexpr auto max_real = std::numeric_limits<T>::max();
+
+    for (auto axis : range(celeritas::Axis::size_))
+    {
+        auto ax = to_int(axis);
+        if (std::isinf((*point)[ax]))
+        {
+            (*point)[ax] = std::copysign(max_real, (*point)[ax]);
+        }
+    }
+}
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Read a bounding box from a JSON file.
+ *
+ * A bounding box is a pair of lower/upper points, with the exception of a
+ * "null" (enclosing no points) bounding box.
+ */
+template<class T>
+void from_json(nlohmann::json const& j, BoundingBox<T>& bbox)
+{
+    if (j.is_null())
+    {
+        // Special case: null bounding box
+        bbox = {};
+        return;
+    }
+
+    CELER_VALIDATE(j.is_array() && j.size() == 2,
+                   << "bounding box must have lower and upper extents");
+
+    // Replace large values (substituted from +-inf) with inf
+    auto lower = j[0].get<Array<T, 3>>();
+    auto upper = j[1].get<Array<T, 3>>();
+    max_to_inf(&lower);
+    max_to_inf(&upper);
+
+    bbox = BoundingBox<T>{lower, upper};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write a bounding box to a JSON file.
+ */
+template<class T>
+void to_json(nlohmann::json& j, BoundingBox<T> const& bbox)
+{
+    if (!bbox)
+    {
+        // Special case: null bounding box
+        j = nullptr;
+        return;
+    }
+
+    // Replace unrepresentable infinities with large values
+    auto lower = bbox.lower();
+    auto upper = bbox.upper();
+    inf_to_max(&lower);
+    inf_to_max(&upper);
+
+    j = nlohmann::json::array({lower, upper});
```

Comment:
is the `nlohman` conventions to transformed `max` into `"null"` ?

Reply:
I think it transforms any unrepresentable number to null: ind, -inf, or nan. JSON itself doesn't allow inf or the like. Replacing with signed max is the only practical solution.

Reply:
> Replacing with signed max is the only practical solution.

Alright, so if I understand correctly the json file now contains literally the numerical value of `std::numeric_limits<T>::max()`, right?

If yes, then do we have a guaranteed that for all values we are using a size (number of bits) guaranteed to be stable on all platforms (eg. `int` or `long` are not).  If not, the produced file would become platform dependent.

Reply:
Correct. I think the potential issue isn't the number of representable bits (these are all IEEE, remember) but slight differences in the text representation of the same value. The produced file can be, but that's the case based on the NLJSON version as well. The important part is whether exporting on A and importing on B changes anything substantial. For ~ulp level differences the answer is that it might disrupt a random number stream or change an exact hash, but nothing should really be different. In this case, we're replacing a representable number `1e308` with an overflow value `inf`. In any place where we don't have an `isinf` these will be effectively equivalent.

The BIH construction *does* have a special case for +-inf, but maybe that threshold should be changed to huge as well? Or huge - epsilon? Or huge * .5 ? (Where does it end???)

Reply:
>  I think the potential issue isn't the number of representable bits

I am confused.  If on platform `A` the size of `T` is 32 bits and the value `inf` ... humm oh right ... this actually in practice only applies to floating points which have stable number of bits, so my point is moot (the point was that `std::numeric_limits<T>::max()` will have completely different values on different platforms for `int` and `long`).

> The BIH construction does have a special case for +-inf, but maybe that threshold should be changed to huge as well? Or huge - epsilon? Or huge * .5 ? (Where does it end???)

The question is how the results varies if `BIH` sees `+inf` vs `huge - epsilon` (essentially the nature of the special casing).

Reply:
Performance will be the only difference; and it may not even be much. (Just having to check the "big" volume more than once potentially.)